File: library/src/main/java/org/incendo/interfaces/InterfacesListeners.java
package org.incendo.interfaces;

import com.github.benmanes.caffeine.cache.Cache;
import com.github.benmanes.caffeine.cache.Caffeine;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.event.Cancellable;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.ClickType;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryCloseEvent;
import org.bukkit.inventory.InventoryHolder;
import org.bukkit.plugin.Plugin;
import org.incendo.interfaces.click.ClickContext;
import org.incendo.interfaces.click.ClickHandler;
import org.incendo.interfaces.click.CompletableClickHandler;
import org.incendo.interfaces.element.CompletedElement;
import org.incendo.interfaces.grid.GridPoint;
import org.incendo.interfaces.interfaces.CloseHandler;
import org.incendo.interfaces.view.AbstractInterfaceView;

import java.util.UUID;
import java.util.concurrent.TimeUnit;
import java.util.logging.Level;

public final class InterfacesListeners implements Listener {

    private static InterfacesListeners instance;

    private static final int OUTSIDE_SLOT = -999;
    private static final int INVENTORY_WIDTH = 9;
    private static final long CLICK_THROTTLE_DURATION = 200L;

    private final Plugin plugin;
    private final Cache<UUID, Object> clickThrottleCache = Caffeine.newBuilder()
        .expireAfterWrite(CLICK_THROTTLE_DURATION, TimeUnit.MILLISECONDS)
        .build();


    private InterfacesListeners(final Plugin plugin) {
        this.plugin = plugin;
        instance = this;
    }

    public static void install(final Plugin plugin) {
        Bukkit.getPluginManager().registerEvents(new InterfacesListeners(plugin), plugin);
    }

    public static InterfacesListeners instance() {
        return instance;
    }

    @EventHandler
    public void onInventoryClose(final InventoryCloseEvent event) {
        InventoryHolder holder = event.getInventory().getHolder();
        AbstractInterfaceView<?, ?> view = this.getInterfaceView(holder);
        if (view == null) {
            return;
        }
        CloseHandler closeHandler = view.backing().closeHandlers().get(event.getReason());
        if (closeHandler != null) {
            closeHandler.handle(event.getReason(), view);
        }
    }

    @EventHandler
    public void onInventoryClick(final InventoryClickEvent event) {
        InventoryHolder holder = event.getInventory().getHolder();
        AbstractInterfaceView<?, ?> view = this.getInterfaceView(holder);
        if (view == null) {
            return;
        }
        GridPoint point = this.getClickPosition(event);
        if (point == null) {
            return;
        }
        this.handleInventoryClick(view, point, event.getClick(), event);
    }

    private AbstractInterfaceView<?, ?> getInterfaceView(final InventoryHolder holder) {
        if (holder instanceof AbstractInterfaceView<?, ?>) {
            return (AbstractInterfaceView<?, ?>) holder;
        }
        return null;
    }

    private GridPoint getClickPosition(final InventoryClickEvent event) {
        int slot = event.getRawSlot();
        if (slot == OUTSIDE_SLOT) {
            return null;
        }
        int y = slot % INVENTORY_WIDTH;
        int x = slot / INVENTORY_WIDTH;
        return GridPoint.at(x, y);
    }

    private void handleInventoryClick(
        final AbstractInterfaceView<?, ?> view,
        final GridPoint point,
        final ClickType clickType,
        final Cancellable cancellable
    ) {
        if (this.shouldCancelClick(view)) {
            cancellable.setCancelled(true);
            return;
        }
        view.setProcessingClick(true);
        try {
            ClickContext context = new ClickContext(view.player(), view, clickType);
            this.processClickPreprocessors(view, context);
            CompletedElement element = view.pane().get(point);
            if (element == null) {
                cancellable.setCancelled(true);
                return;
            }
            this.processElementClick(element, context, cancellable);
        } catch (Exception e) {
            this.plugin.getLogger().log(Level.SEVERE, "Error processing click at " + point, e);
            cancellable.setCancelled(true);
        } finally {
            view.setProcessingClick(false);
        }
    }

    private boolean shouldCancelClick(final AbstractInterfaceView<?, ?> view) {
        return view.isProcessingClick() || this.isClickThrottled(view.player());
    }

    private void processClickPreprocessors(final AbstractInterfaceView<?, ?> view, final ClickContext context) {
        view.backing().clickPreprocessors().forEach(handler -> ClickHandler.process(handler, context));
    }

    private void processElementClick(
        final CompletedElement element,
        final ClickContext context,
        final Cancellable cancellable
    ) {
        CompletableClickHandler clickHandler = ClickHandler.process(element.clickHandler(), context);
        if (!clickHandler.completingLater()) {
            clickHandler.complete();
        }
        cancellable.setCancelled(clickHandler.cancelled());
    }

    private boolean isClickThrottled(final Player player) {
        UUID playerId = player.getUniqueId();
        if (this.clickThrottleCache.getIfPresent(playerId) == null) {
            this.clickThrottleCache.put(playerId, new Object());
            return false;
        }
        return true;
    }

    public Plugin plugin() {
        return this.plugin;
    }

}

File: library/src/main/java/org/incendo/interfaces/InterfacesListeners.kt
package org.incendo.interfaces

import com.github.benmanes.caffeine.cache.Cache
import com.github.benmanes.caffeine.cache.Caffeine
import kotlinx.coroutines.launch
import org.bukkit.Bukkit
import org.bukkit.entity.HumanEntity
import org.bukkit.entity.Player
import org.bukkit.event.Cancellable
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.Action
import org.bukkit.event.inventory.ClickType
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.inventory.InventoryCloseEvent
import org.bukkit.event.inventory.InventoryCloseEvent.Reason
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.event.player.PlayerQuitEvent
import org.bukkit.inventory.EquipmentSlot
import org.bukkit.inventory.InventoryHolder
import org.bukkit.plugin.Plugin
import org.incendo.interfaces.Constants.SCOPE
import org.incendo.interfaces.next.click.ClickContext
import org.incendo.interfaces.click.ClickHandler
import org.incendo.interfaces.click.CompletableClickHandler
import org.incendo.interfaces.next.grid.GridPoint
import org.incendo.interfaces.next.pane.PlayerPane
import org.incendo.interfaces.next.view.AbstractInterfaceView
import org.incendo.interfaces.next.view.InterfaceView
import org.incendo.interfaces.next.view.PlayerInterfaceView
import java.util.*
import java.util.concurrent.TimeUnit

public class InterfacesListeners private constructor() : Listener {

    public companion object {
        /** The current instance for interface listeners class. */
        public lateinit var INSTANCE: org.incendo.interfaces.InterfacesListeners
            private set

        /** Installs interfaces into this plugin. */
        public fun install(plugin: Plugin) {
            require(!org.incendo.interfaces.InterfacesListeners.Companion::INSTANCE.isInitialized) { "Already installed!" }
            org.incendo.interfaces.InterfacesListeners.Companion.INSTANCE = org.incendo.interfaces.InterfacesListeners()
            Bukkit.getPluginManager().registerEvents(org.incendo.interfaces.InterfacesListeners.Companion.INSTANCE, plugin)
        }

        private val VALID_REASON = EnumSet.of(
            Reason.PLAYER,
            Reason.UNKNOWN,
            Reason.PLUGIN
        )

        private val VALID_INTERACT = EnumSet.of(
            Action.LEFT_CLICK_AIR,
            Action.LEFT_CLICK_BLOCK,
            Action.RIGHT_CLICK_AIR,
            Action.RIGHT_CLICK_BLOCK
        )

        private val PLAYER_INVENTORY_RANGE = 0..40
        private const val OUTSIDE_CHEST_INDEX = -999
    }

    private val spamPrevention: Cache<UUID, Unit> = Caffeine.newBuilder()
        .expireAfterWrite(200.toLong(), TimeUnit.MILLISECONDS)
        .build()

    /** A cache of open player interface views, with weak values. */
    private val openPlayerInterfaceViews: Cache<UUID, PlayerInterfaceView> = Caffeine.newBuilder()
        .weakValues()
        .build()

    /** Returns the currently open interface for [playerId]. */
    public fun getOpenInterface(playerId: UUID): PlayerInterfaceView? =
        openPlayerInterfaceViews.getIfPresent(playerId)

    /** Updates the currently open interface for [playerId] to [view]. */
    public fun setOpenInterface(playerId: UUID, view: PlayerInterfaceView?) {
        if (view == null) {
            openPlayerInterfaceViews.invalidate(playerId)
        } else {
            openPlayerInterfaceViews.put(playerId, view)
        }
    }

    @EventHandler
    public fun onClose(event: InventoryCloseEvent) {
        val holder = event.inventory.holder
        val reason = event.reason

        if (holder !is InterfaceView) {
            return
        }

        SCOPE.launch {
            val view = convertHolderToInterfaceView(holder)
            if (view != null) {
                view.backing.closeHandlers[reason]?.invoke(reason, view)
            }

            if (reason !in org.incendo.interfaces.InterfacesListeners.Companion.VALID_REASON) {
                return@launch
            }

            getOpenInterface(event.player.uniqueId)?.open()
        }
    }

    @EventHandler
    public fun onClick(event: InventoryClickEvent) {
        val holder = event.inventory.holder
        val view = convertHolderToInterfaceView(holder) ?: return

        val clickedPoint = clickedPoint(event) ?: return

        handleClick(view, clickedPoint, event.click, event)
    }

    @EventHandler
    public fun onPlayerQuit(event: PlayerQuitEvent) {
        setOpenInterface(event.player.uniqueId, null)
    }

    private fun clickedPoint(event: InventoryClickEvent): GridPoint? {
        // not really sure why this special handling is required,
        // the ordered pane system should solve this but this is the only
        // place where it's become an issue.
        if (event.inventory.holder is Player) {
            val index = event.slot

            if (index !in org.incendo.interfaces.InterfacesListeners.Companion.PLAYER_INVENTORY_RANGE) {
                return null
            }

            val x = index / 9
            val adjustedX = PlayerPane.PANE_ORDERING.indexOf(x)
            return GridPoint(adjustedX, index % 9)
        }

        val index = event.rawSlot

        if (index == org.incendo.interfaces.InterfacesListeners.Companion.OUTSIDE_CHEST_INDEX) {
            return null
        }

        return GridPoint.at(index / 9, index % 9)
    }

    @EventHandler
    public fun onInteract(event: PlayerInteractEvent) {
        if (event.action !in org.incendo.interfaces.InterfacesListeners.Companion.VALID_INTERACT) {
            return
        }
        if (event.hand != EquipmentSlot.HAND) {
            return
        }

        val player = event.player
        val view = getOpenInterface(player.uniqueId) as? AbstractInterfaceView<*, *> ?: return

        val slot = player.inventory.heldItemSlot
        val clickedPoint = GridPoint.at(3, slot)

        val click = convertAction(event.action, player.isSneaking)

        handleClick(view, clickedPoint, click, event)
    }

    /**
     * Converts an inventory holder to an [AbstractInterfaceView] if possible. If the holder is a player
     * their currently open player interface is returned.
     */
    public fun convertHolderToInterfaceView(holder: InventoryHolder?): AbstractInterfaceView<*, *>? {
        if (holder == null) {
            return null
        }

        // If it's an abstract view use that one
        if (holder is AbstractInterfaceView<*, *>) {
            return holder
        }

        // If it's the player's own inventory use the held one
        if (holder is HumanEntity) {
            return getOpenInterface(holder.uniqueId)
        }

        return null
    }

    private fun handleClick(
        view: AbstractInterfaceView<*, *>,
        clickedPoint: GridPoint,
        click: ClickType,
        event: Cancellable
    ) {
        if (view.isProcessingClick || shouldThrottle(view.player)) {
            event.isCancelled = true
            return
        }

        view.isProcessingClick = true

        val clickContext = ClickContext(view.player, view, click)

        view.backing.clickPreprocessors
            .forEach { handler -> org.incendo.interfaces.click.ClickHandler.process(handler, clickContext) }

        val clickHandler = view.pane.getRaw(clickedPoint)
            ?.clickHandler ?: org.incendo.interfaces.click.ClickHandler.ALLOW

        val completedClickHandler = clickHandler
            .run { org.incendo.interfaces.click.CompletableClickHandler().apply { handle(clickContext) } }
            .onComplete { view.isProcessingClick = false }

        if (!completedClickHandler.completingLater) {
            completedClickHandler.complete()
        }

        event.isCancelled = completedClickHandler.cancelled
    }

    private fun convertAction(action: Action, sneaking: Boolean): ClickType {
        if (action.isRightClick) {
            if (sneaking) {
                return ClickType.SHIFT_RIGHT
            }

            return ClickType.RIGHT
        }

        if (action.isLeftClick) {
            if (sneaking) {
                return ClickType.SHIFT_LEFT
            }

            return ClickType.LEFT
        }

        return ClickType.UNKNOWN
    }

    private fun shouldThrottle(player: Player): Boolean =
        if (spamPrevention.getIfPresent(player.uniqueId) == null) {
            spamPrevention.put(player.uniqueId, Unit)
            false
        } else {
            true
        }
}
